src/locale/keymap-util.c

```diff
@@ -780,3 +782,141 @@ int x11_convert_to_vconsole(Context *c) {
 
         return modified;
 }
+
+static
+bool locale_encoding_is_utf8_or_unspecified(const char *locale) {
+        return !strstr(locale, ".") || endswith_no_case(locale, "UTF-8");
+}
+
+bool locale_gen_check_available(void) {
+        if (access("/usr/sbin/locale-gen", X_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether /usr/sbin/locale-gen exists: %m");
+                return false;
+        }
+        if (access("/etc/locale.gen", F_OK) < 0) {
+                if (errno != ENOENT)
+                        log_warning_errno(errno, "Unable to determine whether /etc/locale.gen exists: %m");
+                return false;
+        }
+        return true;
+}
+
+int locale_gen_enable_locale(char *locale[_VARIABLE_LC_MAX]) {
+        _cleanup_fclose_ FILE *fr = NULL, *fw = NULL;
+        _cleanup_(unlink_and_freep) char *temp_path = NULL;
+        _cleanup_free_ char *locale_entry = NULL;
+        bool locale_enabled = false, first_line = false;
+        struct stat st;
+        int r;
+
+        if (isempty(locale[VARIABLE_LANG]))
+                return -EINVAL;
+
+        if (locale_encoding_is_utf8_or_unspecified(locale[VARIABLE_LANG])) {
+                locale_entry = strjoin(locale[VARIABLE_LANG], " UTF-8");
+                if (!locale_entry)
+                        return -ENOMEM;
+        } else
+                return -EINVAL; /* We do not process non-UTF-8 locale */
```

Comment:
EINVAL is such a generic error, which we already use in other contexts heere (just a few lines above, after all). Maybe return a more recognizable error here. Maybe -ENOXEC or so?

Reply:
How about `-ENOTSUP`? That's maybe even more clear (non-UTF-8 locale *could* be supported, but I don't think it's worth doing so and may actually be a disservice. Any experienced user who really wants non-UTF-8 locale could run locale-gen explicitly and manually still though, so not much harm would be done)

Reply:
ENOTSUP is a weird alias, the actual error code is actually EOPNOTSUPP in userspace.

typically EOPNOTSUPP means the operation is not supported, not so much the requested parameters for it, but the lines are blurry.